runtime/runtime/src/verifier.rs

```diff
@@ -124,32 +124,43 @@ pub fn verify_and_charge_tx_ephemeral(
     transaction_cost: &TransactionCost,
     block_height: Option<BlockHeight>,
     current_protocol_version: ProtocolVersion,
+    ephemeral_state: Option<(Account, AccessKey)>,
 ) -> Result<VerificationResult, InvalidTxError> {
-    let _span = tracing::debug_span!(target: "runtime", "verify_and_charge_transaction").entered();
+    let _span = tracing::debug_span!(target: "runtime", "verify_and_charge_tx_ephemeral").entered();
 
     let TransactionCost { gas_burnt, gas_remaining, receipt_gas_price, total_cost, burnt_amount } =
         *transaction_cost;
 
     let tx = validated_tx.to_tx();
     let signer_id = tx.signer_id();
 
-    let mut signer = match get_account(state_update, signer_id)? {
-        Some(signer) => signer,
+    // TODO(miloserdow): Move the logic of calling `get_account`/`get_access_key` to the caller(s).
+    let (mut signer, mut access_key) = match ephemeral_state {
+        Some((ephemeral_signer, ephemeral_access_key)) => (ephemeral_signer, ephemeral_access_key),
```

Comment:
I think it is worth it to add a `debug_assert` that the cached values match what the actual lookups would return.  It might help catch bugs and helps with readability.

Reply:
`debug_assert` here would be difficult if for no other reason than that it would require adding in writes back to storage for each transaction processed.